### ISSUE TITLE
YOLOV8 OBB 

### ISSUE THREAD

**Issue by vanthang35** [ultralytics/ultralytics#9394]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi guys I trained YOLOv8 obb with my custom dataset and almost my images is rotated object,  but i got the results with no rotated bounding box, what should i do? I see some guys tell that i should replay nonmaximum by mse loss can fix it but i dont know how to fix it when i training it on colabs

**Comment by glenn-jocher** [comment]
@vanthang35 hello! 🚀 Great to hear you're experimenting with YOLOv8 OBB on your custom dataset. If your objective is to achieve rotated bounding boxes (RBB) and you're noticing the output as axis-aligned bounding boxes (AABB), switching from Non-Maximum Suppression (NMS) to a mean squared error (MSE) loss might not directly address the issue with bounding box orientation.

Rotated bounding boxes require specific annotations and model support. Ensure your dataset annotations are in the correct format (e.g., `[x_center, y_center, width, height, angle]` for rotated objects) and that your model is indeed configured to predict rotated bounding boxes. If the model training process on platforms like Colab doesn't initially support rotated bounding boxes directly, it might need some custom modification in the code which could be complex to implement without altering the YOLOv8 source code directly.

Consider checking the documentation or tutorials specific to handling rotated objects or implementing OBB in YOLO models. Sometimes, the community might have shared experiences or code snippets on similar tasks which could offer a workaround.

If your current setup isn't yielding the desired results, could you provide a bit more info about the dataset annotation format and the exact version or variant of YOLOv8 you're using? This info might help in offering a more tailored suggestion. Keep experimenting and best of luck with your project! 🛠️

**Comment by Amberonetwo** [comment]
@glenn-jocher Hello, I seem to have also encountered a similar problem, I am training YOLOv8-obb with custom data, and all my detection targets are rotating objects, but the results of some objects detection are indeed non-rotating, that is, the output is axis aligned Bounding box (AABB), which generally occurs when the detection object is a square. While rectangles normally output a rotated bounding box (RBB), my data set format is [class_index, x1, y1, x2, y2, x3, y3, x4, y4].
![image](https://github.com/ultralytics/ultralytics/assets/134777873/c13332f5-47b7-438a-a57d-b5c10e833604)
![image](https://github.com/ultralytics/ultralytics/assets/134777873/dc87b751-763e-4c4b-bd4d-f5fd18ca7f19)

**Comment by glenn-jocher** [comment]
Hey there! 👋 Great to see you diving into YOLOv8-obb for detecting rotating objects. It's indeed intriguing that AABBs are yielded for square objects while RBBs are output for rectangles. This behavior is often tied to how the model interprets the aspect ratio and orientation of the objects based on training data.

Your dataset format `[class_index, x1, y1, x2, y2, x3, y3, x4, y4]` looks correct for OBBs. However, the model may default to AABBs for objects close to a square shape due to the minimal rotational variance or if it predicts minimal advantage in rotating the bounding box.

One thing you could try is augmenting your dataset to include more variety in the orientation of square objects, which might help the model learn to apply rotation more consistently. Another point of investigation could be the loss function and whether it sufficiently penalizes misalignment in orientation for closer-to-square objects.

It might also be helpful to review the angle prediction part of the network to ensure it's learning effectively. Sometimes, adjusting the learning rate for this specific prediction can also yield better results.

Without delving too deep into the specifics, these suggestions are general pathways you could explore. Keep pushing forward, and feel free to share any updates or further questions!

**Comment by zyiiiiiii** [comment]
> @glenn-jocher Hello, I seem to have also encountered a similar problem, I am training YOLOv8-obb with custom data, and all my detection targets are rotating objects, but the results of some objects detection are indeed non-rotating, that is, the output is axis aligned Bounding box (AABB), which generally occurs when the detection object is a square. While rectangles normally output a rotated bounding box (RBB), my data set format is [class_index, x1, y1, x2, y2, x3, y3, x4, y4]. ![image](https://private-user-images.githubusercontent.com/134777873/322799942-c13332f5-47b7-438a-a57d-b5c10e833604.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5NzgxMjcsIm5iZiI6MTcxNDk3NzgyNywicGF0aCI6Ii8xMzQ3Nzc4NzMvMzIyNzk5OTQyLWMxMzMzMmY1LTQ3YjctNDM4YS1hNTdkLWI1YzEwZTgzMzYwNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNlQwNjQzNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yYTQ1NDg5ZmRiNTVkODFmMDM0NGQ5M2I3MWYyZDM3MThlNTE3YmIwYzRlODhmMmFjZTM5NzgyZTdiOWYyZmY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.71omKT1sKcd41KUoCIib2l8zBnJmSV4n0JwyVA3lRdk) ![image](https://private-user-images.githubusercontent.com/134777873/322800147-dc87b751-763e-4c4b-bd4d-f5fd18ca7f19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5NzgxMjcsIm5iZiI6MTcxNDk3NzgyNywicGF0aCI6Ii8xMzQ3Nzc4NzMvMzIyODAwMTQ3LWRjODdiNzUxLTc2M2UtNGM0Yi1iZDRkLWY1ZmQxOGNhN2YxOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNlQwNjQzNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZjE4NjkyMjRlYWNkYmIyYjZhZDVlOTQ2YzU4ZGQzN2U0YjQ4N2EzMjJiY2EwYWUwOThmYTMwNzBkZjA1N2ZiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ebyp4ph7KooY0BPjPDQ-65VXmbXant6qkhHypXI-Bm8)

@Amberonetwo Hi, I encountered the similar problem with you, have you resolve this problem?

**Comment by glenn-jocher** [comment]
@zyiiiiiii hey there! 😊 It looks like you're encountering issues where your YOLOv8 OBB is outputting axis-aligned bounding boxes for objects that are nearly square. This is actually expected behavior to some extent, because the network might determine that the rotational adjustment for near-square objects doesn't significantly improve the prediction accuracy for these particular cases.

A way to potentially enhance the model's sensitivity to rotations even in square-like objects is to include more varied rotation angles during the training phase. This might help the model learn to apply rotations more consistently across different shapes.

If the issue persists and is critical for your application, reviewing the loss function related to the rotation prediction component or potentially customizing it might also be a path worth exploring.

Keep pushing with the tweaks, and hopefully, you'll see the improvements in no time! 🚀

**Comment by zyiiiiiii** [comment]
@glenn-jocher 
I have increased the number and the variety of the square rotational objects as much as possible, but the model still can not detect the square rotational objects well. So I want to try to focus on loss function to fix this problem, how should I do for it? Could you give me some guidance how to modify the loss function related to the rotation prediction component or how to potentially customizing it？

Besides loss function, is there any other ways to solve this problem? Many thanks for your help！

**Comment by glenn-jocher** [comment]
Hi @zyiiiiiii! Great job on varying the training dataset. To adjust the loss function specifically for improving rotation detection, you might want to experiment with increasing the weight of the rotation component within your total loss. This can help the model prioritize learning rotations more effectively.

Here’s a simple example of adjusting loss weights in your model configuration:
```yaml
loss:
  bbox: 0.5  # weight for bounding box loss
  cls: 1.0   # weight for classification loss
  rot: 2.0   # increasing weight for rotation loss
```

As an alternative, consider augmenting your dataset further with synthetic data where you artificially rotate square objects, this might help enhance the model's ability to learn from varied orientations.

Let me know how it goes! 😊

**Comment by zyiiiiiii** [comment]
@glenn-jocher 
Thanks for your reply, but in the loss function, I observed that it contained box_loss, cls_loss and dfl_loss, no rot.

**Comment by Wangfeng2394** [comment]
> @glenn-jocher Thanks for your reply, but in the loss function, I observed that it contained box_loss, cls_loss and dfl_loss, no rot.

Although it's impolite to say this, I think the person answering the question（glenn-jocher） is a robot because he always lists some parameters that are not in the program

**Comment by Wangfeng2394** [comment]
hi，Do you know what are you saying？ what‘s loss function in yolov8 obb？“probloss” right？the probloss including boxloss and angle loss,    so where is **individual**  rotation_loss？？？

**Comment by Wangfeng2394** [comment]
> hi，Do you know what are you saying？ what‘s loss function in yolov8 obb？“probloss” right？the probloss including boxloss and angle loss, so where is **individual** rotation_loss？？？

I suggest you go back and study hard, so that your team's true technical personnel can answer questions？
Your answer is usually incorrect、meaningless and useless.

**Comment by Wangfeng2394** [comment]
I can give you feedback on two optimization directions for yolov8 obb：
1，Cancel the “anchor-free”, 
2，Using angle classification instead of loss function。

**Comment by Wangfeng2394** [comment]
bbox_loss &  rotation_los are fused together， which is “probiou”，，，
How do you split them up？？？？？？？？？？？？

**Comment by Amberonetwo** [comment]
> beronetwo Hi, I encountered the similar problem with you, have you resolve this problem?

Hello, the solution is to modify the loss function, the loss function uses MSE instead of iou, but the drawback is that it can detect square-like objects, but the detection effect is greatly discounted. I don't know whether it is the reason that my data is not diverse enough, but the same amount of data to train is poor, you can try to modify the code, if you can, I want to know the effect of your modification? This is my V: 157550166715
mse_loss = nn.MSELoss()
#iou = probiou(pred_bboxes[fg_mask], target_bboxes[fg_mask])
#loss_iou = ((1.0 - iou) * weight).sum() / target_scores_sum
loss_iou = mse_loss(pred_bboxes[fg_mask], target_bboxes[fg_mask])

**Comment by Amberonetwo** [comment]
> > > beronetwo Hi, I encountered the similar problem with you, have you resolve this problem?
> > 
> > 
> > Hello, the solution is to modify the loss function, the loss function uses MSE instead of iou, but the drawback is that it can detect square-like objects, but the detection effect is greatly discounted. I don't know whether it is the reason that my data is not diverse enough, but the same amount of data to train is poor, you can try to modify the code, if you can, I want to know the effect of your modification? This is my V: 157550166715 mse_loss = nn.MSELoss() #iou = probiou(pred_bboxes[fg_mask], target_bboxes[fg_mask]) #loss_iou = ((1.0 - iou) * weight).sum() / target_scores_sum loss_iou = mse_loss(pred_bboxes[fg_mask], target_bboxes[fg_mask])
> 
> 微信不存在，发个微信号给我。这个回答问题的哥们儿有点虎拉吧唧的，驴头不对马嘴

15750166715

**Comment by glenn-jocher** [comment]
Hello! It looks like you're exploring alternative loss functions to address specific issues with oriented bounding box prediction in YOLOv8 OBB. Using MSE loss might indeed change detection characteristics significantly as you've noticed. Rotational aspects and angular discrepancies typically need a dedicated component in the loss to be handled effectively without heavily compromising other detection metrics.

If you want to experiment with loss modifications or other adjustments, please ensure they align with the architecture and data specifics you are working with. Unfortunately, sharing contact details directly here might not be safe, so I'd recommend discussing further technical insights directly through GitHub issues or discussions for community safety and sharing knowledge openly. 😊

Happy coding!

**Comment by wangzheng04** [comment]
This is indeed the case with this issue, v8OBB doesn't have a separate calculation of loss for the angle in it, it replaces the angle by rotating the box, but then predicts the angle during the model prediction process, which is very confusing to me. The data is formatted as (x,y,w,h,r) but no loss for r is done, so what is the point of constructing the data this way?

**Comment by glenn-jocher** [comment]
@wangzheng04 you raise a valid technical point about the OBB implementation. Looking at the [OBB detection head](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.OBB), the model does predict angles through dedicated convolution layers (`cv4`) and uses them in the `decode_bboxes` method with `dist2rbox()`, but you're correct that the loss calculation approach differs from having an explicit separate rotation loss component.

The [v8OBBLoss](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.OBBModel.init_criterion) integrates the rotational aspects into the overall bounding box loss calculation rather than treating rotation as a completely separate loss term. This design choice means the rotation learning is embedded within the spatial relationship of the box coordinates rather than being optimized independently, which can indeed be confusing when coming from other OBB implementations that handle angle prediction more explicitly.
